src/frontend/src/optimizer/plan_node/logical_expand.rs

```diff
@@ -0,0 +1,227 @@
+        Self::new(input, expanded_keys).into()
+    }
+
+    fn derive_schema(input_schema: &Schema) -> Schema {
+        let mut fields = input_schema.clone().into_fields();
+        fields.push(Field::with_name(DataType::Int64, "flag"));
+        Schema::new(fields)
+    }
+
+    pub fn expanded_keys(&self) -> &Vec<Vec<usize>> {
+        &self.expanded_keys
+    }
+}
+
+impl PlanTreeNodeUnary for LogicalExpand {
+    fn input(&self) -> PlanRef {
+        self.input.clone()
+    }
+
+    fn clone_with_input(&self, input: PlanRef) -> Self {
+        Self::new(input, self.expanded_keys.clone())
+    }
+
+    #[must_use]
+    fn rewrite_with_input(
+        &self,
+        input: PlanRef,
+        input_col_change: ColIndexMapping,
+    ) -> (Self, ColIndexMapping) {
+        let mut expanded_keys = self.expanded_keys.clone();
+        for key in expanded_keys.iter_mut().flat_map(|r| r.iter_mut()) {
+            *key = input_col_change.map(*key);
+        }
+        let (mut map, new_input_col_num) = input_col_change.into_parts();
+        assert_eq!(new_input_col_num, input.schema().len());
+        map.push(Some(new_input_col_num));
+
+        (Self::new(input, expanded_keys), ColIndexMapping::new(map))
+    }
+}
+
+impl_plan_tree_node_for_unary! {LogicalExpand}
+
+impl fmt::Display for LogicalExpand {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "LogicalExpand {{ expanded_keys: {:#?} }}",
+            self.expanded_keys
+        )
+    }
+}
+
+impl ColPrunable for LogicalExpand {
+    fn prune_col(&self, required_cols: &[usize]) -> PlanRef {
+        let pos_of_flag = self.input.schema().len();
+        let input_required_cols = required_cols
+            .iter()
+            .copied()
+            .filter(|i| *i != pos_of_flag)
+            .collect_vec();
+        let new_input = self.input.prune_col(&input_required_cols);
+
+        // `input_required_cols` should be a subset of `expanded_keys`
+        let input_change = ColIndexMapping::with_remaining_columns(
+            &input_required_cols,
+            self.input.schema().len(),
+        );
+        // Filter those unneeded `expanded_keys`.
+        let expanded_keys = self
+            .expanded_keys
+            .iter()
+            .filter_map(|keys| {
+                let keys = keys
+                    .iter()
+                    .filter_map(|key| input_change.try_map(*key))
+                    .collect_vec();
+                if keys.is_empty() {
+                    None
+                } else {
+                    Some(keys)
+                }
+            })
+            .collect_vec();
+        LogicalExpand::create(new_input, expanded_keys)
+    }
+}
+
+impl PredicatePushdown for LogicalExpand {
+    fn predicate_pushdown(&self, predicate: Condition) -> PlanRef {
+        let new_input = self.input.predicate_pushdown(predicate);
```

Comment:
What about the sub-predicate that references `flag` column?

Reply:
Yes, I think here are some potential correctness issues. e.g. what will happen if there is a predicate `a == NULL` and a is in subsets, we are not supposed to push it down.